<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px"> 

## Lab: SQL Practice

_Author: Matt Brems_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [1]:
# Import pandas and create_engine from `sqlalchemy`

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data
engine = \
create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [2]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(*)
FROM stores

"""

stores = pd.read_sql_query(sql, engine)
stores

count
0   1973

**Prompt 1 Answer**:
There are 1973 rows in the stores table.

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [3]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT DISTINCT name
FROM stores
ORDER BY name DESC

"""

name_stores = pd.read_sql_query(sql, engine)
name_stores.head(4)

name
0        Z's Quickbreak
1  Zapf's Pronto Market
2        Yr Dollar Stop
3         Xpress Liquor

**Prompt 2 Answer**: The fourth sort by name stores is Xpress Liquor.

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [4]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT btl_price, bottle_qty, category_name
FROM sales
WHERE category_name = 'IMPORTED VODKA'
ORDER BY vendor ASC
LIMIT 3;
"""

sales = pd.read_sql_query(sql, engine)

In [5]:
sales

btl_price  bottle_qty   category_name
0     $8.99          12  IMPORTED VODKA
1    $28.50           6  IMPORTED VODKA
2    $28.50           1  IMPORTED VODKA

**Prompt 3 Answer**: 
- First transactions :	price $8.99, 12 bottles

- Second transactions :	price $28.50, 3 bottles

- Third transactions :	price $28.50, 3 bottles

### Prompt 4

What is the total amount of liquor sold in February 2015?

In [6]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT SUM(total)::MONEY
FROM sales
WHERE date BETWEEN '2-1-2015' AND '2-28-2015';
"""

date_total = pd.read_sql_query(sql, engine)

In [7]:
date_total

sum
0  $21,295,350.55

**Prompt 4 Answer**: 
- The total amount of liquor sold in February 2015 is 21,295,350.55 $

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [8]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT vendor, COUNT(*) AS number_of_transactions,
SUM(bottle_qty) AS number_of_bottles,
SUM(total) AS revenue
FROM sales 
WHERE category_name = 'IMPORTED VODKA'
GROUP BY 1
ORDER BY COUNT(*) DESC;
"""
vendor_trans = pd.read_sql_query(sql, engine)

In [9]:
vendor_trans.head()

vendor  number_of_transactions  \
0  Pernod Ricard USA/Austin Nichols                   42338   
1               Bacardi U.S.A. Inc.                   26441   
2                   Diageo Americas                   15725   
3   Constellation Wine Company Inc.                   12565   
4                   Jim Beam Brands                    5850   

   number_of_bottles     revenue  
0             403450  7880365.88  
1             236047  5771242.07  
2             152038  3346425.42  
3             116053  1895005.51  
4             104119  1807097.23

**Prompt 5 Answer**: 
- The most transactions vendor is Pernod Ricard USA/Austin Nichols as 42,338 transactions.

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [10]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price - state_btl_cost AS markup 
FROM sales
"""

makeup = pd.read_sql_query(sql, engine)

In [11]:
makeup["markup"] = makeup["markup"].replace({'\$': '', ',': ''}, regex = True)

In [12]:
makeup["markup"].astype(float).mean()
# Average cost amount of extra money charged about 4.87 $ 

4.807856588060757

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [13]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT store, AVG( (btl_price - state_btl_cost) / state_btl_cost * 100 ) AS "Percentage Markup"
FROM sales
GROUP BY 1
ORDER BY 2 DESC;
"""

per_markup= pd.read_sql_query(sql, engine)

In [14]:
per_markup.head(1)

store  Percentage Markup
0   4013          52.547124

**Prompt 7 Answer**: 
- The store 4103 is the largest average markup that have 52.54%.

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [15]:
# Use a SQL query to find the answer to the above prompt.

sql = sql = """
SELECT r.store, r.name, r.store_address, AVG( (btl_price - state_btl_cost) / state_btl_cost * 100 ) AS "Percentage Markup"
FROM sales as o
    LEFT JOIN stores as r
    ON o.store = r.store
GROUP BY r.store
ORDER BY 4 DESC
LIMIT 5;
"""

store_markup = pd.read_sql_query(sql, engine)

In [16]:
store_markup[["name"]]

name
0      Wal-Mart 0841 / Tipton
1  Wal-Mart 1546 / Iowa Falls
2  Wal-Mart 1683 / Shenandoah
3    Wal-Mart 3394 / Atlantic
4   Wal-Mart 2935 / Knoxville

**Prompt 8 Answer**:

The names of the five stores are
- 	Wal-Mart 0841 / Tipton
-	Wal-Mart 1546 / Iowa Falls
-	Wal-Mart 1683 / Shenandoah
-	Wal-Mart 3394 / Atlantic
-	Wal-Mart 2935 / Knoxville

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery).

In [17]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT r.county, (SUM(bottle_qty) / r.population) AS "bottles of liquor", 
ROUND(AVG(re.bottle_size),1) AS "average bottle size"
FROM sales as o
    JOIN counties AS r ON o.county = r.county
    JOIN products AS re ON o.item = re.item_no
GROUP BY 1
ORDER BY 2 DESC
LIMIT 9
"""

country = pd.read_sql_query(sql, engine)

In [18]:
country["county"].values

array(['Dickinson', 'Polk', 'Black Hawk', 'Cerro Gordo', 'Scott',
       'Johnson', 'Linn', 'Pottawattamie', 'Kossuth'], dtype=object)

In [19]:
country.sort_values(by="average bottle size")

county  bottles of liquor  average bottle size
2     Black Hawk                 15                862.9
4          Scott                 13                871.6
6           Linn                 12                886.3
1           Polk                 15                892.6
7  Pottawattamie                 11                893.7
5        Johnson                 13                895.4
0      Dickinson                 22                974.1
3    Cerro Gordo                 14                986.4
8        Kossuth                 11               1041.0

**Prompt 9 Answer**: 
- Nine counties sell the highest number of bottles of liquor per capita 
    'Dickinson', 'Polk', 'Black Hawk', 'Cerro Gordo', 'Scott',
     'Johnson', 'Linn', 'Pottawattamie' and 'Kossuth'
- County sells (on average) the largest bottle size is Black Hawk	

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [20]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT (btl_price * bottle_qty) AS calculated_total, total
FROM sales
WHERE CAST ((btl_price * bottle_qty) AS NUMERIC) <> total
"""

sales = pd.read_sql_query(sql, engine)

In [21]:
sales

Empty DataFrame
Columns: [calculated_total, total]
Index: []

**Prompt 10 Answer**: 0

## Additional problems

In [21]:
sql = """
SELECT *
FROM 
    stores
LIMIT
    1
"""
stores = pd.read_sql_query(sql, engine)

In [24]:
sql = """
SELECT *
FROM 
    sales
LIMIT
    1
"""
sales = pd.read_sql_query(sql, engine)

In [25]:
sql = """
SELECT *
FROM 
    counties
LIMIT
    1
"""
counties = pd.read_sql_query(sql, engine)

In [26]:
sql = """
SELECT *
FROM 
    products
LIMIT
    1
"""
products = pd.read_sql_query(sql, engine)

In [29]:
stores.T

0
store                                                       2106
name                                 Hillstreet News and Tobacco
store_status                                                   A
store_address  2217 College\nCedar Falls, IA 506130000\n(42.5...
address_info                                                None

In [30]:
sales.T

0
date                            2014-11-13 00:00:00
convenience_store                              None
store                                          2635
county_number                                    82
county                                        Scott
category                                    1032200
category_name                 IMPORTED VODKA - MISC
vendor_no                                       370
vendor             Pernod Ricard USA/Austin Nichols
item                                          34036
description                           Absolut Pears
pack                                             12
liter_size                                      750
state_btl_cost                               $11.49
btl_price                                    $17.24
bottle_qty                                        4
total                                         68.96

In [31]:
counties.T

0
county      Adair
population   7682

In [32]:
products.T

0
item_no                                       904616
category_name     MISC. AMERICAN CORDIALS & LIQUEURS
item_description              Travis Hasse Apple Pie
vendor                                           305
vendor_name                                  Mhw Ltd
bottle_size                                      750
pack                                              12
inner_pack                                         1
age                                             None
proof                                             40
list_date                        2009-02-11 00:00:00
upc                                             None
scc                                             None
bottle_price                                   $9.77
shelf_price                                    14.66
case_cost                                     117.22

---

- 1. How many total products are in the Products table? 

In [5]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT 
    COUNT(DISTINCT item_no)
FROM 
    products
"""

product = pd.read_sql_query(sql, engine)

In [7]:
product
# 9975 products

count
0   9977

---

-  2. Who are the top most diverse vendors (i.e. they have the highest number of distinct products)? How many different products do they have?

In [19]:
sql = """
SELECT 
    vendor,
    vendor_name,
    COUNT(DISTINCT item_no) As distinct_products
FROM 
    products
GROUP BY
    vendor, vendor_name
ORDER BY
    3 DESC
LIMIT
    5
"""

div_product = pd.read_sql_query(sql, engine)
div_product
# Jim Beam Brands 

vendor                       vendor_name  distinct_products
0      65                   Jim Beam Brands                925
1     260                   Diageo Americas                907
2     370  Pernod Ricard Usa/austin Nichols                599
3     885         Yahara Bay Distillers Inc                579
4     259     Heaven Hill Distilleries Inc.                388

---

- 3. Which products sell the best by total number of unit sales? 

(Hint: you will need to find a way of doing a cross-join between the sales and product tables.)

In [67]:
sql = """
SELECT 
    r.item_no,
    r.category_name,
    SUM(o.bottle_qty) As total_unit
FROM sales as o
    JOIN products AS r ON o.item = r.item_no
GROUP BY
    1,2
ORDER BY
    3 DESC
LIMIT
    5
"""

total_unit = pd.read_sql_query(sql, engine)
total_unit
# Products sell the best by total number of unit sales is 11788 of item_no and CANADIAN WHISKIES type.

item_no      category_name  total_unit
0    11788  CANADIAN WHISKIES      828393
1    36308     80 PROOF VODKA      500070
2    43337         SPICED RUM      483621
3    36904     80 PROOF VODKA      393203
4    35317     80 PROOF VODKA      288425

---

- 4. Which products sell the best by total dollar value of sales?

In [74]:
sql = """
SELECT 
    r.item_no,
    r.category_name,
    SUM(o.total)::MONEY As total_dollar
FROM sales as o
    JOIN products AS r ON o.item = r.item_no
GROUP BY
    1, 2
ORDER BY
    3 DESC
LIMIT
    5
"""

total_dollar = pd.read_sql_query(sql, engine)
total_dollar
# Products sell the best by total dollar value of sales is 11788 of item_no and CANADIAN WHISKIES type.

item_no       category_name    total_dollar
0    11788   CANADIAN WHISKIES  $12,863,376.81
1    43337          SPICED RUM   $8,523,370.55
2    26827  TENNESSEE WHISKIES   $6,668,863.94
3    11297   CANADIAN WHISKIES   $6,231,299.41
4    36308      80 PROOF VODKA   $5,380,753.20

---

- 5 What are the top 10 categories of liquor sold based on the total amount of sales revenue?

In [73]:
sql = """
SELECT 
    r.category_name,
    SUM(o.total)::MONEY As total_dollar
FROM sales as o
    JOIN products AS r ON o.item = r.item_no
GROUP BY
    1
ORDER BY
    2 DESC
LIMIT
    10
"""

top_dollar = pd.read_sql_query(sql, engine)
top_dollar
# ['CANADIAN WHISKIES', '80 PROOF VODKA', 'SPICED RUM',
# 'IMPORTED VODKA', 'TEQUILA', 'STRAIGHT BOURBON WHISKIES',
# 'WHISKEY LIQUEUR', 'TENNESSEE WHISKIES',
# 'PUERTO RICO & VIRGIN ISLANDS RUM', 'BLENDED WHISKIES']

category_name    total_dollar
0                 CANADIAN WHISKIES  $48,059,037.63
1                    80 PROOF VODKA  $48,040,023.20
2                        SPICED RUM  $31,586,392.99
3                    IMPORTED VODKA  $23,884,662.13
4                           TEQUILA  $21,413,525.94
5         STRAIGHT BOURBON WHISKIES  $20,904,099.16
6                   WHISKEY LIQUEUR  $19,346,811.30
7                TENNESSEE WHISKIES  $17,600,801.58
8  PUERTO RICO & VIRGIN ISLANDS RUM  $12,729,072.76
9                  BLENDED WHISKIES  $12,018,198.55

In [77]:
top_dollar["category_name"].values

array(['CANADIAN WHISKIES', '80 PROOF VODKA', 'SPICED RUM',
       'IMPORTED VODKA', 'TEQUILA', 'STRAIGHT BOURBON WHISKIES',
       'WHISKEY LIQUEUR', 'TENNESSEE WHISKIES',
       'PUERTO RICO & VIRGIN ISLANDS RUM', 'BLENDED WHISKIES'],
      dtype=object)

---